opencog/atoms/core/ScopeLink.cc

```diff
@@ -313,29 +358,23 @@ ContentHash ScopeLink::term_hash(const Handle& h,
 		const Variables& vees = sco->get_variables();
 		for (const Handle& v : vees.varseq) bound_vars.insert(v);
 	}
-
-	// Prevent mixing for UnorderedLinks. The `mixer` var will be zero
-	// for UnorderedLinks. The problem is that two UnorderdLinks might
-	// be alpha-equivalent, but have their atoms presented in a
-	// different order. Thus, the hash must be computed in a purely
-	// commutative fashion: using only addition, so as to never create
-	// any entropy, until the end.
-	//
-	// XXX As discussed in issue #1176, a better fix would be to
-	// compute the individual term_hashes first, then sort them,
-	// and then mix them!  This provides the desired qualities:
-	// different unordered links can be directly compared, and also
-	// have good mixing/avalanching properties. The code below
-	// only allows for compare; it fails to mix.
-	//
-	bool is_ordered = not classserver().isA(t, UNORDERED_LINK);
-	ContentHash mixer = (ContentHash) is_ordered;
+	// As discussed in issue #1176, compute the individual term_hashes
+	// first, then sort them, and then mix them!  This provides the
+	// desired qualities: different unordered links can be directly
+	// compared, and also have good mixing/avalanching properties.
 	ContentHash hsh = ((1UL<<8) - 59) * t;
+	std::vector<ContentHash> hash_vec;
 	for (const Handle& ho: h->getOutgoingSet())
 	{
-		hsh += mixer * (hsh<<5) + term_hash(ho, bound_vars, quotation);
+		hash_vec.push_back(term_hash(ho, bound_vars, quotation));
+	}
+	// hash_vec should be sorted only for unordered links
+	if (classserver().isA(t, UNORDERED_LINK)) {
+		std::sort(hash_vec.begin(), hash_vec.end());
```

Comment:
Specifically, the actual sort order used should be identical to that in the constructors in `atoms/core/UnorderedLink.cc`

Reply:
See also the `unorder()` method in `atoms/core/FreeLink.cc`. 

*If* you can guarantee the the ctors for these various link types have already run, by the time that you compute the hash, then you are home-free.  However, I am not completely sure if they have.  In the past, many of these links types did not have ctors, and so the sorting was done ad-hoc, in the atomspace.   However, I think this has now been fixed. I think.  I just did a `grep -r sort *` and do not see any problem areas.  Still, caution is urged...

Reply:
@linas said
> Specifically, the actual sort order used should be identical to that in the constructors in 
atoms/core/UnorderedLink.cc

I don't see why it could need to be the same
Are you sure? Why would it need to be? It's just to calculate the hash. And if variables are renamed to account for alpha-equivalence, it likely isn't the same already, right?

Reply:
OK, I see what how the `UnorderedLink` constructor affects `Link::compute_hash`. However, I don't think it has any impact on how this code work, because the hash obtained by `ScopeLink::term_hash` is simply different from the hash obtained from `Link::compute_hash`, and the later isn't used at all.

That is said, it is indeed good for performances if both return the same order as then it makes the sorting in `ScopeLink::term_hash` cheaper (cause it's already sorted), but that should be all.